### PR TITLE
Minimize the amount of calls into `event.bp`

### PR DESCRIPTION
**Pull request by thewhitegoatcb**
Resolves https://github.com/actboy168/lua-debug/issues/259
Needs more testing for non jit lua as I'm not an expert.
The performance boost is over 100x for awaiting breakpoints.
Tweaks are done by @fesily suggestions

### CODE REVIEW

File: src/luadebug/rdebug_hookmgr.cpp [hunk 1]
```diff
@@ -463,17 +464,19 @@ struct hookmgr {
         default:
             return;
         }
-        push_callback(L);
-        luadebug::debughost::set(L, hL);
```
Comment (actboy168) [inline src/luadebug/rdebug_hookmgr.cpp:467]: Here should always call back into lua. If not, then it shouldn't have gotten here.

**Review by actboy168 — CHANGES_REQUESTED** [review]

File: src/luadebug/rdebug_hookmgr.cpp (hunk 1 above)
Comment (thewhitegoatcb) [reply src/luadebug/rdebug_hookmgr.cpp:467]: You are right.
Rechecking the code it seems like there's a bug in luajit's implementation, in specific on how returning from C call is being handled, have to find a better solution

File: src/luadebug/rdebug_hookmgr.cpp [hunk 2]
```diff
@@ -463,17 +464,19 @@ struct hookmgr {
...
+            luadebug::debughost::set(L, hL);
             luadbg_pushstring(L, "step");
             luadbg_pushinteger(L, ar->currentline);
             if (luadbg_pcall(L, 2, 0, 0) != LUADBG_OK) {
                 luadbg_pop(L, 1);
                 return;
             }
         }
-        else {
+        else if (break_mask & LUA_MASKLINE) {
```
Comment (thewhitegoatcb) [inline src/luadebug/rdebug_hookmgr.cpp:477]: Possibly replace `if (break_mask & LUA_MASKLINE)` replace it with
```C
#ifdef LUAJIT_VERSION 
if (break_mask & LUA_MASKLINE) 
#endif
```

Comment (actboy168) [reply src/luadebug/rdebug_hookmgr.cpp:477]: @fesily

**Comment by fesily** [comment]
First of all, one thing is clear, this pull should be a generic one and not a luajit specific one.

Secondly, the code now has to call event.bp because lua-debug is not sure if the proto and source information of the current lua function has been read? @actboy168

**Comment by thewhitegoatcb** [comment]
In my understanding, The issue arises due to an assumption that `case LUA_HOOKLINE:` is only executed when we found a breakpoint on the previous hook call.
That would be wrong in luajit's case, because we are doing this 
```lua
#ifdef LUAJIT_VERSION
            last_hook_call_in_c = !lua_isluafunc(hL, ar);
            if (last_hook_call_in_c) {
                thread_mask |= LUA_MASKLINE;
                updatehookmask(hL);

                if (update_mask)
                    update_hook(hL);
            }
#endif
```
in `LUA_HOOKCALL, LUA_HOOKTAILCALL, LUA_HOOKTAILRET`, hence we getting lots of `event.bp` on C func exits. I don't think it's the intended use of it, as it wouldn't happen in standard Lua (I didn't test, but the code seems to indicate that).

**Comment by fesily** [comment]
> > > First of all, one thing is clear, this pull should be a generic one and not a luajit specific one.
> > > Secondly, the code now has to call event.bp because lua-debug is not sure if the proto and source information of the current lua function has been read? @actboy168
> > 
> > 
> > wouldn't that be done in `newproto` that's called by `break_new`?
> 
> So this question needs to be answered by actboy168. As I understand it bp events are only valid in the bpmap::status::Break state.

You found the core problem, I do not see. It should be here to deal with it.
I was calling the c function because when I stepped in, the hit position would not be the second line of the function

**Comment by fesily** [comment]
@thewhitegoatcb 
```c
void full_hook(lua_State* hL, lua_Debug* ar) {
        const auto is_step_mode = [this]() { return (step_mask & LUA_MASKLINE) && (!stepL || stepL == hL); };
        switch (ar->event) {
        case LUA_HOOKLINE:
#ifdef LUAJIT_VERSION
            if (last_hook_call_in_c) {
#    if defined(LUA_HOOKTHREAD)
                thread_mask &= (~LUA_MASKTHREAD);
#    endif
                updatehookmask(hL);
                last_hook_call_in_c = false;
                if (break_mask)
                    break_hook_return_luajit(hL, ar);
            }
            if (stepL == hL) {
                if (step_mask & LUA_MASKRET) {
                    step_hook_line(hL, ar);
                }
            }
            if (!is_step_mode() && !(break_mask & LUA_MASKLINE))
                return;
#endif
            break;
```
try the code

**Comment by fesily** [comment]
@thewhitegoatcb the code maybe better
```diff
diff --git a/src/luadebug/rdebug_hookmgr.cpp b/src/luadebug/rdebug_hookmgr.cpp
index 5331e3c0..3750c682 100644
--- a/src/luadebug/rdebug_hookmgr.cpp
+++ b/src/luadebug/rdebug_hookmgr.cpp
@@ -393,13 +393,28 @@ struct hookmgr {
                 updatehookmask(hL);
                 last_hook_call_in_c = false;
             }
+            else
+#endif
+                break;
+        case LUA_HOOKRET:
+            if (update_mask) {
+                update_hook(hL);
+            }
+            if (break_mask & LUA_MASKRET) {
+                break_hook_return(hL, ar);
+            }
             if (stepL == hL) {
                 if (step_mask & LUA_MASKRET) {
-                    step_hook_line(hL, ar);
+                    step_hook_return(hL, ar);
                 }
             }
+#if LUA_VERSION_NUM >= 504 || defined(LUAJIT_VERSION)
+            else if (step_mask & LUA_MASKLINE) {
+                // step in
+                break;
+            }
 #endif
-            break;
+            return;
         case LUA_HOOKCALL:
 #if LUA_VERSION_NUM >= 502
         case LUA_HOOKTAILCALL:
@@ -422,28 +437,6 @@ struct hookmgr {
             if (last_hook_call_in_c) {
                 thread_mask |= LUA_MASKLINE;
                 updatehookmask(hL);
-
-                if (update_mask)
-                    update_hook(hL);
-            }
-#endif
-            return;
-        case LUA_HOOKRET:
-            if (update_mask) {
-                update_hook(hL);
-            }
-            if (break_mask & LUA_MASKRET) {
-                break_hook_return(hL, ar);
-            }
-            if (stepL == hL) {
-                if (step_mask & LUA_MASKRET) {
-                    step_hook_return(hL, ar);
-                }
-            }
-#if LUA_VERSION_NUM >= 504
-            else if (step_mask & LUA_MASKLINE) {
-                // step in
-                break;
             }
 #endif
             return;
```

**Comment by thewhitegoatcb** [comment]
@fesily works perfectly, great job
minor quirk
if there's a code like this
```lua
c_func() and c_func()
```
and I put a breakpoint and hit continue after trigger, it will trigger again
forgot to test if it happens with normal lua functions too
just a minor thing

**Comment by fesily** [comment]
> @fesily works perfectly, great job minor quirk if there's a code like this
> 
> ```lua
> c_func() and c_func()
> ```
> 
> and I put a breakpoint and hit continue after trigger, it will trigger again forgot to test if it happens with normal lua functions too just a minor thing

This temporarily do not deal with it, you can modify the pull

**Comment by thewhitegoatcb** [comment]
@fesily I think I have made a mistake while testing.
Tested again yesterday, your proposed last fix doesn't seem to fix performance.
but the previous one with `is_step_mode` does work

**Comment by fesily** [comment]
> @fesily I think I have made a mistake while testing. Tested again yesterday, your proposed last fix doesn't seem to fix performance. but the previous one with `is_step_mode` does work

As long as it works

**Comment by fesily** [comment]
nogc64 this function you should submit a new pull

**Comment by thewhitegoatcb** [comment]
yes, my bad I should have moved it to a different branch

**Comment by fesily** [comment]
@actboy168 合并吧,这样就可以了

**Pull request merged**